File: local-console-ui/src/app/core/module/edgeapp.ts
/**
 * Copyright 2024 Sony Semiconductor Solutions Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { DTDLReqInfoV2, DTDLResInfoV2 } from './dtdl';

// Module: edge app
//  - configuration: TODO: Spec unknown
//  - state
export interface EdgeAppModuleConfigurationV2 {
  edge_app?: EdgeAppModuleEdgeAppV2;
}

// Module: edge app
// configuration:
// - edge_app
export interface EdgeAppModuleEdgeAppV2 {
  req_info?: DTDLReqInfoV2;
  res_info?: DTDLResInfoV2;
  common_settings?: EdgeAppModuleEdgeAppCommonSettingsV2;
  custom_settings?: EdgeAppModuleEdgeAppCustomSettingsV2;
}

export interface EdgeAppModuleEdgeAppCustomSettingsV2 {
  ai_models?: {
    one_pass_model?: {
      parameters?: {
        threshold?: number;
        input_width?: number;
        input_height?: number;
        max_detections?: number;
        dnn_output_detections?: number;
      };
      ai_model_bundle_id?: string;
    };
  };
}

export interface EdgeAppModuleEdgeAppCommonSettingsV2 {
  log_level?: number;
  pq_settings?: EdgeAppModuleEdgeAppPQSettingsV2;
  port_settings?: EdgeAppModuleEdgeAppPortSettingsV2;
  process_state?: number;
  codec_settings?: {
    format?: number;
  };
  upload_interval?: number;
  inference_settings?: {
    number_of_iterations?: number;
  };
  number_of_inference_per_message?: number;
}

export interface EdgeAppModuleEdgeAppPortSettingsV2 {
  metadata?: {
    path?: string;
    method?: number;
    enabled?: boolean;
    endpoint?: string;
    storage_name?: string;
  };
  input_tensor?: {
    path?: string;
    method?: number;
    enabled?: boolean;
    endpoint?: string;
    storage_name?: string;
  };
}

export interface EdgeAppModuleEdgeAppPQSettingsV2 {
  frame_rate?: {
    num?: number;
    denom?: number;
  };
  digital_zoom?: number;
  auto_exposure?: {
    max_gain?: number;
    convergence_speed?: number;
    max_exposure_time?: number;
    min_exposure_time?: number;
  };
  exposure_mode?: number;
  image_cropping?: {
    top?: number;
    left?: number;
    width?: number;
    height?: number;
  };
  image_rotation?: number;
  ev_compensation?: number;
  manual_exposure?: {
    gain?: number;
    exposure_time?: number;
  };
  camera_image_flip?: {
    flip_vertical?: number;
    flip_horizontal?: number;
  };
  camera_image_size?: {
    width?: number;
    height?: number;
    scaling_policy?: number;
  };
  auto_white_balance?: {
    convergence_speed?: number;
  };
  white_balance_mode?: number;
  ae_anti_flicker_mode?: number;
  manual_white_balance_gain?: {
    red?: number;
    blue?: number;
  };
  manual_white_balance_preset?: {
    color_temperature?: number;
  };
}

export interface EdgeAppModuleStateV2 {
  edge_app?: EdgeAppModuleEdgeAppV2;
}

File: local-console/src/local_console/core/camera/v2/components/edge_app.py
# Copyright 2024 Sony Semiconductor Solutions Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0
from enum import Enum
from typing import Any

from local_console.core.camera.v2.components.req_res_info import ReqInfo
from local_console.core.camera.v2.components.req_res_info import ResInfo
from pydantic import BaseModel


class ProcessState(Enum):
    STOPPED = 1
    RUNNING = 2


class UploadMethod(Enum):
    MQTT_TELEMETRY = 0
    BLOB_STORAGE = 1
    HTTP_STORAGE = 2


class CodecSettingsSpec(BaseModel):
    format: int | None = None


class CameraImageSizeSpec(BaseModel):
    width: int | None = None
    height: int | None = None
    scaling_policy: int | None = None


class ManualWhiteBalanceGainSpec(BaseModel):
    red: int | float | None = None
    blue: int | float | None = None


class CameraImageFlipSpec(BaseModel):
    flip_vertical: int | None = None
    flip_horizontal: int | None = None


class FrameRateSpec(BaseModel):
    num: int | None = None
    denom: int | None = None


class AutoExposureSpec(BaseModel):
    max_gain: int | float | None = None
    convergence_speed: int | float | None = None
    max_exposure_time: int | float | None = None
    min_exposure_time: int | float | None = None


class AutoWhiteBalanceSpec(BaseModel):
    convergence_speed: int | float | None = None


class ManualExposureSpec(BaseModel):
    gain: int | float | None = None
    exposure_time: int | float | None = None


class ImageCroppingSpec(BaseModel):
    top: int | None = None
    left: int | None = None
    width: int | None = None
    height: int | None = None


class ManualWhiteBalancePresetSpec(BaseModel):
    color_temperature: int | float | None = None


class PQSettings(BaseModel):
    frame_rate: FrameRateSpec | None = None
    digital_zoom: int | float | None = None
    auto_exposure: AutoExposureSpec | None = None
    exposure_mode: int | float | None = None
    image_cropping: ImageCroppingSpec | None = None
    image_rotation: int | float | None = None
    ev_compensation: int | float | None = None
    manual_exposure: ManualExposureSpec | None = None
    camera_image_flip: CameraImageFlipSpec | None = None
    camera_image_size: CameraImageSizeSpec | None = None
    auto_white_balance: AutoWhiteBalanceSpec | None = None
    white_balance_mode: int | float | None = None
    ae_anti_flicker_mode: int | float | None = None
    manual_white_balance_gain: ManualWhiteBalanceGainSpec | None = None
    manual_white_balance_preset: ManualWhiteBalancePresetSpec | None = None


class UploadSpec(BaseModel):
    enabled: bool = False
    method: UploadMethod | None = None
    endpoint: str | None = None
    path: str | None = None
    storage_name: str | None = None


class EdgeAppPortSettings(BaseModel):
    metadata: UploadSpec | None = None
    input_tensor: UploadSpec | None = None


class InferenceSettingsSpec(BaseModel):
    number_of_iterations: int | None = None


class EdgeAppCommonSettings(BaseModel):
    log_level: int | None = None
    port_settings: EdgeAppPortSettings | None = None
    pq_settings: PQSettings | None = None
    process_state: ProcessState | None = None
    codec_settings: CodecSettingsSpec | None = None
    upload_interval: int | None = None
    inference_settings: InferenceSettingsSpec | None = None
    number_of_inference_per_message: int | None = None


class EdgeAppSpec(BaseModel):
    req_info: ReqInfo | None = None
    res_info: ResInfo | None = None
    common_settings: EdgeAppCommonSettings | None = None
    custom_settings: Any | None = None


APP_CONFIG_KEY = "edge_app"
